include/crm/common/messages_internal.h

```diff
@@ -119,4 +120,28 @@ pcmk__request_origin(const pcmk__request_t *request)
     }
 }
 
+/*!
+ * \internal
+ * \brief Store a deep copy of an XML node in a named wrapper under a parent
+ *
+ * Result:
+ * * \p msg contains a new element named \p name
+ * * The \p name element contains a copy of \p src
+ *
+ * The copy can then be retrieved via <tt>get_message_xml(msg, name)</tt>. This
+ * is a convenience function for adding arbitrary XML to a message (for example,
+ * an IPC or CPG message).
+ *
+ * \param[in,out] msg   Parent to create the wrapper element under
+ * \param[in]     name  Name of wrapper element
+ * \param[in]     src   XML to copy
+ */
+static inline void
+pcmk__message_add_xml(xmlNode *msg, const char *name, xmlNode *src)
+{
+    // No way to retrieve the copy later if msg == NULL
+    CRM_CHECK(msg != NULL, return);
+    pcmk__xml_copy(pcmk__xe_create(msg, name), src);
```

Comment:
> obscures what's being done

Kind of yeah. It looks a little odd when expanded FWIW, with the payload inside two layers of nesting. But that might still be better.

> rather than message-specific

yeah that always got me

> If we were starting from scratch

Yeah I think I would prefer that. I'm not looking forward to making that change though, if I'm ever the person to do it. Especially with rolling-upgrade compatibility in `pacemaker-based`.